homeassistant/config_entries.py

```diff
@@ -1169,6 +1175,52 @@ def _data_to_save(self) -> dict[str, list[dict[str, Any]]]:
         """Return data to save."""
         return {"entries": [entry.as_dict() for entry in self._entries.values()]}
 
+    async def async_remove_device(self, entry_id: str, device_id: str) -> bool:
+        """Remove a device."""
+        if (config_entry := self.async_get_entry(entry_id)) is None:
+            return False
+
+        if not config_entry.supports_remove_device:
+            return False
+
+        dev_reg = device_registry.async_get(self.hass)
+        ent_reg = entity_registry.async_get(self.hass)
+
+        if (device_entry := dev_reg.async_get(device_id)) is None:
+            return False
+
+        if entry_id not in device_entry.config_entries:
+            return False
+
+        try:
+            integration = await loader.async_get_integration(
+                self.hass, config_entry.domain
+            )
+        except loader.IntegrationNotFound:
+            return False
+        try:
+            component = integration.get_component()
+        except ImportError:
+            return False
+
+        # Remove device entities which belong to the config entry
+        entity_entries = entity_registry.async_entries_for_device(
+            ent_reg, device_id, True
+        )
+        for entity_entry in entity_entries:
+            if entity_entry.config_entry_id == entry_id:
+                ent_reg.async_remove(entity_entry.entity_id)
+
+        # Remove the config entry from the device, this will also remove the device
+        # if it was the only config entry.
+        dev_reg.async_update_device(device_id, remove_config_entry_id=entry_id)
+
+        # Inform the config entry in case it needs to take additional actions
+        await component.async_config_entry_removed_from_device(
```

Comment:
The idea is that the integration may need to take additional actions, for example update some remote device or service.
Should we call this first, and allow the integration to veto the removal?

Reply:
Yes, we should call it first and only remove the entities and device entries if the return value is True.